smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/DocumentationPreprocessor.kt

```diff
@@ -21,17 +26,226 @@ class DocumentationPreprocessor : KotlinIntegration {
+                    text += if (node.parent()?.nodeName() == "body") "\n\n" else ""
+                }
+                "code", "pre" -> {
+                    text += PREFORMAT_MARKER
+                }
+                "b", "strong" -> {
+                    text += BOLD_MARKER
+                }
+                "i", "em" -> {
+                    text += ITALIC_MARKER
+                }
+                "li", "fullname", "dt" -> {
+                    text += "\n"
+                }
+            }
+        }
+
+        private fun writeBufferedAnchor() {
+            // Model docs will sometimes contain an anchor without the href. At that point there's no real way of
+            // knowing to what it refers, nor can we guarantee a valid link just by bracketing it.
+            text += if (bufferedAnchorHref != "") {
+                "[$bufferedAnchorText]($bufferedAnchorHref)"
+            } else {
+                bufferedAnchorText
+            }
+
+            bufferedAnchorHref = ""
+            bufferedAnchorText = ""
+        }
+
+        private fun sublistIndent() {
+            listPrefix += SUBLIST_INDENT
+        }
+
+        private fun sublistDedent() {
+            listPrefix = listPrefix.dropLast(SUBLIST_INDENT.length)
+        }
+    }
+}
+
+/**
+ * Jsoup will preserve newlines between elements as blank text nodes. These have zero bearing on the content of the
+ * document to begin with and only serve to complicate traversal.
+ */
+private fun Node.stripBlankTextNodes() {
+    if (this is TextNode && this.isBlank) {
+        this.remove()
+        return
+    }
+
+    val childNodes = this.childNodes()
+    if (childNodes.isNotEmpty()) {
+        childNodes.forEach(Node::stripBlankTextNodes)
+    }
+}
+
+private fun Node.hasAncestor(predicate: (Node) -> Boolean): Boolean {
+    if (!this.hasParent()) return false
+
+    val parent = this.parent() as Node
+    return predicate(parent) || parent.hasAncestor(predicate)
+}
+
+private fun Node.isList() =
+    this.nodeName() == "ul" || this.nodeName() == "ol"
+
+private fun TextNode.markdownText() =
+    this.text()
+        // Replace square brackets with escaped equivalents so that they are not rendered as invalid Markdown
+        // links.
+        .replace("[", "&#91;")
+        .replace("]", "&#93;")
+
+/**
+ * Operates on all substrings that fall within the provided section delimiters. Returns a new string where all
+ * substrings enclosed as specified have been modified according to the provided transform.
+ */
+private fun String.applyWithin(start: String, end: String, transform: (String) -> String): String {
+    val substringStart = this.indexOf(start) + start.length
+    if (substringStart == -1) return this
+
+    val substringEnd = this.indexOf(end, substringStart)
+    if (substringEnd == -1) return this
+
+    val stringToTransform = this.substring(substringStart, substringEnd)
+    return this.substring(0, substringStart) + transform(stringToTransform) + end +
```

Comment:
This actually might be easier as a [Regex](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/-regex/) which has a `replace` function that will replace all occurrences.

Reply:
Ya regex may be difficult to get right but something like:

```
<code>(.*?)<\/code>
```
With option to enable `.` to also match newline characters would probably work.

e.g.

```kotlin
val startToken = Regex.escape(start)
val endToken = Regex.escape(end)
val re = Regex("$startToken(.*?)$endToken", setOf(Regex.DOT_MATCHES_ALL))
...
```

Your approach is fine as well though, it's up to you.